dvc/repo/add.py

```diff
@@ -38,6 +40,26 @@ def add(
         raise RecursiveAddingWhileUsingFilename()
 
     targets = ensure_list(targets)
+
+    invalid_opt = None
+    if to_remote:
+        message = "{option} can't be used with --to-remote"
+        if len(targets) != 1:
+            invalid_opt = "multiple targets"
+        elif no_commit:
+            invalid_opt = "--no-commit option"
+        elif recursive:
+            invalid_opt = "--recursive option"
+    else:
+        message = "{option} can't be used without --to-remote"
+        if kwargs.get("out"):
+            invalid_opt = "--out"
```

Comment:
Similar Q. Why introduce this limitation? Might as well enable `add --out` for any file along the way? Someone could find it useful perhaps e.g. add the same data twice (with different resulting names) to do different transformations (an alternative to branching) — not sure.

Reply:
> add the same data twice

doesn't work (outputs overlap) - it'll complain

Reply:
Hmm I don't see this. If the output paths are (given explicitly) different, where's the overlap?

Maybe my example isn't great though, because it's equivalent to making copies of the data and `dvc add` them all (which is trivial and we already support).

Reply:
okay, yep, I'm not exactly right here. If `--out` controls path data will be located within the project, then yes it can be done probably without causing an overlap (but it'll duplicate data potentially, or rename it in the source? not sure about the semantics here). I wouldn't complicate it for now anyway, let's again wait for someone asking this. Since otherwise we'll have to think at least about semantics, and it looks a a bit like a different feature.

Reply:
It's a different feature indeed. Out of scope for this PR

The only thing is that `add --out` by itself seems to have a meaning but will be forbidden (for now). Not a big deal unless someone complains though

BTW @isidentical maybe some of these restrictions should be handled by printing a warning and ignoring the flag, instead of stopping and throwing an error? Idk, up to you.

Reply:
> maybe some of these restrictions should be handled by printing a warning and ignoring the flag, instead of stopping and throwing an error? Idk, up to you.

I'd personally choose explicit errors rather than some warnings, though I am open to that if it is more convenient for users.

Reply:
OK. Well, if you're not sure maybe check what DVC does in places where similar limitations already exist. See `repro --recursive/all-pipelines/downstream`, `run --always-changed`, and `check-ignore --all` — those are all the options with documented restrictions (but maybe there's more).

Reply:
+1 for being strict - errors by default :)